layer/executables/packager.py

```diff
@@ -0,0 +1,153 @@
+import inspect
+import os
+import pickle  # nosec
+import shutil  # nosec
+import sys
+import tempfile
+import zipapp
+from pathlib import Path
+from typing import Any, Callable, List, Optional
+
+import cloudpickle  # type: ignore
+
+
+def package_function(
+    function: Callable[..., Any],
+    resources: Optional[List[Path]] = None,
+    pip_dependencies: Optional[List[str]] = None,
+    output_dir: Optional[Path] = None,
+) -> Path:
+    """Packages layer function as a Python executable."""
+
+    if not inspect.isfunction(function) or function.__name__ == "<lambda>":
+        raise ValueError("function must be a function")
+
+    with tempfile.TemporaryDirectory() as source_dir:
+        source = Path(source_dir)
+
+        requirements_path = source / "requirements.txt"
+        with open(requirements_path, mode="w", encoding="utf8") as requirements:
+            requirements.write(
+                "\n".join(_runtime_requirements + (pip_dependencies or []))
+            )
+
+        main_path = source / "__main__.py"
+        with open(main_path, mode="w", encoding="utf8") as main:
+            main.write(_loader_source())
+
+        if resources:
+            _package_resources(source, resources)
+
+        function_path = source / "function"
+        with open(function_path, mode="wb") as function_:
+            # register to pickle by value to ensure unpickling works anywhere
+            cloudpickle.register_pickle_by_value(sys.modules[function.__module__])
```

Comment:
Are we sure about pickling by value, I know it wasn't introduced on this PR but based on the discussion on channel it wasn't a winner either.
It would be more in line with previous edition of the function pickling and we don't need to explicitly include cloudpickle as it should've been installed with layer. I think that is a requirement introduced by pickle by value somehow.

Reply:
Not sure about  pickling by value either, but will remove, as less code is always simpler.

Regarding the `cloudpickle` dependency, I think it is better to keep it there explicitly to have a closed loop in serialising deserialising.

Reply:
> Not sure about pickling by value either, but will remove, as less code is always simpler.

After digging into the docs, I think we should keep `register_pickle_by_value` call to have the executables less dependant on the external environments. In fact, removing `cloudpickle.register_pickle_by_value` fails all the tests in `test/unit/executables/test_packager` referencing `func_simple` and `func_with_dependencies`.

Reply:
maybe expand the comment with more info for posterity?